File: Lua/iterators.lua
-- Iterators em Lua geralmente são escritos com closures
--
-- Exemplo com arrays que, diferente de ipairs(), retorna só o elemento, e
-- quando acabar a lista retorna nil:
list_iter = function(t)
    local i = 0
    local n = table.getn(t)
    return function()
        i = i + 1
        if i <= n then return t[i] end
    end
end

-- Como o for chama seu iterator enquanto esse não retornar nil, podemos usá-lo
-- normalmente:
for i in list_iter({1, 2, 3}) do
    print(i)
end

File: Rust/iterators.rs
/*
 * Quanto a iterators, são igualmente métodos de primeira-classe. São lazy, ou seja, só são
 * evaluados quando necessário.
 */

// A trait Iterator é definida assim:

pub trait Iterator {
    type Item;

    fn next(&self) -> Option<Self::Item>;
}

/*
 * Quando um for roda um iterator, chama o método Iterator.next até que retorne None. Caso
 * queiramos iterar "manualmente", podemos inclusive chamar o método next diretamente:
 */
fn main() {
    let list = vec![1, 2, 3];
    let mut iter = list.iter();
    let next_item = iter.next();
    // Perceba que precisamos declarar iter como mutável pois o método next muda seu estado. No
    // caso de um for isso não é necessário pois o for se apropria do iterator e torna-o mutável.

    // Iterator também possuem métodos definidos pela stdlib. Por exemplo, sum() e map():
    let sum = iter.sum();
    // map produz outro iterator:
   let sum2 = iter.map(|x| x + 1).sum();
}
